File: apps/envoygo/test/services/library_read_fetch_test.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';
import 'package:envoygo/models/library_read.dart';
import 'package:envoygo/services/content_hash.dart';
import 'package:envoygo/services/library_read_fetch.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('fetchLibraryContent', () {
    test('returns ok body for single-shot read', () async {
      final result = await fetchLibraryContent(
        ({
          required String targetOwnerId,
          required String path,
          Map<String, int>? range,
          String? ifNoneMatch,
          int? timeoutMs,
        }) async =>
            LibraryReadResult(
              peerOwnerId: targetOwnerId,
              libp2pPeerId: 'p',
              status: 'ok',
              body: '# Hi',
              contentType: 'text/markdown',
              byteLength: 4,
              latencyMs: 1,
            ),
        targetOwnerId: 'envoy:owner:a',
        path: 'hi.md',
      );
      expect(result.status, 'ok');
      expect(result.body, '# Hi');
      expect(result.isText, isTrue);
    });

    test('assembles too_large via range chunks', () async {
      final total = libraryReadChunkBytes + 10;
      final full = List.filled(total, 0x41); // 'A'
      final result = await fetchLibraryContent(
        ({
          required String targetOwnerId,
          required String path,
          Map<String, int>? range,
          String? ifNoneMatch,
          int? timeoutMs,
        }) async {
          if (range == null) {
            return LibraryReadResult(
              peerOwnerId: targetOwnerId,
              libp2pPeerId: 'p',
              status: 'too_large',
              contentType: 'text/plain',
              byteLength: total,
              latencyMs: 1,
            );
          }
          final start = range['start']!;
          final end = range['end']!;
          final slice = full.sublist(start, end + 1);
          return LibraryReadResult(
            peerOwnerId: targetOwnerId,
            libp2pPeerId: 'p',
            status: 'ok',
            body: base64Encode(slice),
            contentType: 'text/plain',
            byteLength: slice.length,
            latencyMs: 1,
          );
        },
        targetOwnerId: 'envoy:owner:a',
        path: 'big.txt',
      );
      expect(result.status, 'ok');
      expect(result.body, 'A' * total);
      expect(result.byteLength, total);
    });

    test('restores body from cache on not_modified', () async {
      const cache = BrowserFetchCacheEntry(
        body: '# Cached',
        contentType: 'text/markdown',
        contentHash: 'abc',
        etag: 'etag123',
        byteLength: 8,
        isText: true,
      );
      final result = await fetchLibraryContent(
        ({
          required String targetOwnerId,
          required String path,
          Map<String, int>? range,
          String? ifNoneMatch,
          int? timeoutMs,
        }) async {
          expect(ifNoneMatch, 'etag123');
          return LibraryReadResult(
            peerOwnerId: targetOwnerId,
            libp2pPeerId: 'p',
            status: 'not_modified',
            etag: 'etag123',
            contentHash: 'abc',
            contentType: 'text/markdown',
            byteLength: 8,
            latencyMs: 1,
          );
        },
        targetOwnerId: 'envoy:owner:a',
        path: 'hi.md',
        cache: cache,
        revalidate: true,
      );
      expect(result.status, 'ok');
      expect(result.fromCache, isTrue);
      expect(result.body, '# Cached');
    });

    test('errors when not_modified without cache', () async {
      final result = await fetchLibraryContent(
        ({
          required String targetOwnerId,
          required String path,
          Map<String, int>? range,
          String? ifNoneMatch,
          int? timeoutMs,
        }) async =>
            LibraryReadResult(
              peerOwnerId: targetOwnerId,
              libp2pPeerId: 'p',
              status: 'not_modified',
              etag: 'x',
              latencyMs: 1,
            ),
        targetOwnerId: 'envoy:owner:a',
        path: 'hi.md',
        revalidate: true,
      );
      expect(result.status, 'error');
      expect(result.error, contains('not_modified'));
    });

    test('propagates forbidden', () async {
      final result = await fetchLibraryContent(
        ({
          required String targetOwnerId,
          required String path,
          Map<String, int>? range,
          String? ifNoneMatch,
          int? timeoutMs,
        }) async =>
            LibraryReadResult(
              peerOwnerId: targetOwnerId,
              libp2pPeerId: 'p',
              status: 'forbidden',
              latencyMs: 1,
            ),
        targetOwnerId: 'envoy:owner:a',
        path: 'secret.md',
      );
      expect(result.status, 'forbidden');
    });

    test('errors when range chunk fails', () async {
      final result = await fetchLibraryContent(
        ({
          required String targetOwnerId,
          required String path,
          Map<String, int>? range,
          String? ifNoneMatch,
          int? timeoutMs,
        }) async {
          if (range == null) {
            return LibraryReadResult(
              peerOwnerId: targetOwnerId,
              libp2pPeerId: 'p',
              status: 'too_large',
              contentType: 'text/plain',
              byteLength: libraryReadChunkBytes + 5,
              latencyMs: 1,
            );
          }
          return LibraryReadResult(
            peerOwnerId: targetOwnerId,
            libp2pPeerId: 'p',
            status: 'not_found',
            latencyMs: 1,
          );
        },
        targetOwnerId: 'envoy:owner:a',
        path: 'big.txt',
      );
      expect(result.status, 'error');
      expect(result.error, contains('range fetch failed'));
    });
  });

  group('verifyContentHash', () {
    test('accepts matching utf8 hash', () {
      const body = '# Hello';
      final hash = sha256.convert(utf8.encode(body)).toString();
      expect(
        verifyContentHash(
          body: body,
          contentType: 'text/markdown',
          expectedHash: hash,
        ),
        isTrue,
      );
    });

    test('rejects mismatch', () {
      expect(
        verifyContentHash(
          body: '# Hello',
          contentType: 'text/markdown',
          expectedHash: '0' * 64,
        ),
        isFalse,
      );
    });

    test('skips when expectedHash absent', () {
      expect(
        verifyContentHash(
          body: 'x',
          contentType: 'text/plain',
          expectedHash: null,
        ),
        isTrue,
      );
    });
  });
}

File: apps/social/test/library-read-fetch.test.ts
import { describe, expect, it, vi } from "vitest";
import {
  baseMimeType,
  fetchLibraryContent,
  LIBRARY_READ_CHUNK_BYTES,
} from "../src/lib/library-read-fetch.js";
import type { LibraryReadResult } from "@envoymesh/api";

function okResult(partial: Partial<LibraryReadResult>): LibraryReadResult {
  return {
    peerOwnerId: "envoy:owner:a",
    libp2pPeerId: "peer",
    latencyMs: 1,
    status: "ok",
    ...partial,
  };
}

describe("baseMimeType", () => {
  it("strips charset parameters", () => {
    expect(baseMimeType("text/markdown; charset=utf-8")).toBe("text/markdown");
    expect(baseMimeType("text/html; charset=utf-8")).toBe("text/html");
    expect(baseMimeType("TEXT/Markdown")).toBe("text/markdown");
  });
});

describe("fetchLibraryContent", () => {
  it("returns a normal ok response", async () => {
    const libraryRead = vi.fn(async () =>
      okResult({
        body: "# Hi",
        contentType: "text/markdown",
        contentHash: "abc",
        etag: "abc".slice(0, 3),
        byteLength: 4,
      }),
    );
    const result = await fetchLibraryContent(libraryRead, {
      targetOwnerId: "envoy:owner:a",
      path: "hello.md",
    });
    expect(result.status).toBe("ok");
    expect(result.body).toBe("# Hi");
    expect(result.isText).toBe(true);
  });

  it("treats text/*; charset=utf-8 as text", async () => {
    const libraryRead = vi.fn(async () =>
      okResult({
        body: "# Hi",
        contentType: "text/markdown; charset=utf-8",
        contentHash: "abc",
        byteLength: 4,
      }),
    );
    const result = await fetchLibraryContent(libraryRead, {
      targetOwnerId: "envoy:owner:a",
      path: "hello.md",
    });
    expect(result.status).toBe("ok");
    expect(result.isText).toBe(true);
  });

  it("assembles base64 range chunks on too_large", async () => {
    const total = LIBRARY_READ_CHUNK_BYTES + 100;
    const full = Buffer.alloc(total, 0x41); // 'A'
    const fullB64 = full.toString("base64");
    const libraryRead = vi.fn(async (params: { range?: { start: number; end: number } }) => {
      if (!params.range) {
        return okResult({
          status: "too_large",
          byteLength: total,
          contentType: "application/pdf",
          contentHash: "deadbeef",
          etag: "deadbeef",
        }) as LibraryReadResult;
      }
      const slice = full.subarray(params.range.start, params.range.end + 1);
      return okResult({
        body: slice.toString("base64"),
        contentType: "application/pdf",
        contentHash: "deadbeef",
        etag: "deadbeef",
        range: { start: params.range.start, end: params.range.end, total },
        byteLength: slice.length,
      });
    });

    const result = await fetchLibraryContent(libraryRead, {
      targetOwnerId: "envoy:owner:a",
      path: "big.pdf",
    });
    expect(result.status).toBe("ok");
    expect(result.body).toBe(fullB64);
    expect(result.isText).toBe(false);
    expect(libraryRead.mock.calls.length).toBeGreaterThan(1);
  });

  it("returns cached body on not_modified when revalidating", async () => {
    const libraryRead = vi.fn(async () =>
      okResult({
        status: "not_modified",
        etag: "etag1",
        contentHash: "hash1",
        contentType: "text/markdown",
        byteLength: 5,
      }) as LibraryReadResult,
    );
    const result = await fetchLibraryContent(libraryRead, {
      targetOwnerId: "envoy:owner:a",
      path: "hello.md",
      revalidate: true,
      cache: {
        body: "# cached",
        contentType: "text/markdown",
        contentHash: "hash1",
        etag: "etag1",
        byteLength: 8,
        isText: true,
      },
    });
    expect(result.status).toBe("ok");
    expect(result.fromCache).toBe(true);
    expect(result.body).toBe("# cached");
    expect(libraryRead).toHaveBeenCalledWith(
      expect.objectContaining({ ifNoneMatch: "etag1" }),
    );
  });
});
